kyo-core/shared/src/main/scala/kyo/Channel.scala

```diff
@@ -8,334 +9,295 @@ import scala.annotation.tailrec
-                                            p.safe.get
-                                        case Present(v) =>
-                                            v
-                            finally
-                                flush()
-                        }
-
-                    def takeFiber(using frame: Frame) =
-                        IO.Unsafe {
-                            try
-                                if u.closed() then
-                                    throw closedException
-                                else
-                                    u.poll() match
-                                        case Absent =>
-                                            val p = Promise.Unsafe.init[Nothing, A]()
-                                            takes.add(p)
-                                            p.safe
-                                        case Present(v) =>
-                                            Fiber.success(v)
-                            finally
-                                flush()
+    def init[A](capacity: Int, access: Access = Access.MultiProducerMultiConsumer)(using Frame): Channel[A] < IO =
+        IO.Unsafe(Unsafe.init(capacity, access))
+
+    /** WARNING: Low-level API meant for integrations, libraries, and performance-sensitive code. See AllowUnsafe for more details. */
+    abstract class Unsafe[A]:
+        def capacity: Int
+        def size()(using AllowUnsafe): Result[Closed, Int]
+
+        def offer(value: A)(using AllowUnsafe): Result[Closed, Boolean]
+        def poll()(using AllowUnsafe): Result[Closed, Maybe[A]]
+
+        def putFiber(value: A)(using AllowUnsafe): Fiber.Unsafe[Closed, Unit]
+        def takeFiber()(using AllowUnsafe): Fiber.Unsafe[Closed, A]
+
+        def drain()(using AllowUnsafe): Result[Closed, Seq[A]]
+        def close()(using Frame, AllowUnsafe): Maybe[Seq[A]]
+
+        def empty()(using AllowUnsafe): Result[Closed, Boolean]
+        def full()(using AllowUnsafe): Result[Closed, Boolean]
+        def closed()(using AllowUnsafe): Boolean
+
+        def safe: Channel[A] = this
+    end Unsafe
+
+    /** WARNING: Low-level API meant for integrations, libraries, and performance-sensitive code. See AllowUnsafe for more details. */
+    object Unsafe:
+        def init[A](
+            _capacity: Int,
+            access: Access = Access.MultiProducerMultiConsumer
+        )(using initFrame: Frame, allow: AllowUnsafe): Unsafe[A] =
+            new Unsafe[A]:
+                import AllowUnsafe.embrace.danger
+
+                val queue = Queue.Unsafe.init[A](_capacity, access)
+                val takes = new MpmcUnboundedXaddArrayQueue[Promise.Unsafe[Closed, A]](8)
+                val puts  = new MpmcUnboundedXaddArrayQueue[(A, Promise.Unsafe[Closed, Unit])](8)
+
+                def capacity = _capacity
+
+                def size()(using AllowUnsafe) = queue.size()
+
+                def offer(value: A)(using AllowUnsafe) =
```

Comment:
I've optimized these methods to only call `flush` if the operation was able to make a change, otherwise it isn't necessary.